File: hosts/vscode-plugin/src/ui/WebguiStaticServer.ts
import * as http from "http";
import * as fs from "fs";
import * as path from "path";
import { logger } from "../globals";

const MIME: Record<string, string> = {
  ".html": "text/html; charset=utf-8",
  ".js": "application/javascript; charset=utf-8",
  ".mjs": "application/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".ico": "image/x-icon",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
  ".ttf": "font/ttf",
  ".wasm": "application/wasm",
  ".map": "application/json",
};

/**
 * Lightweight HTTP server that serves the embedded webgui-dist files
 * under the `/app/` prefix and injects `window.__OPENCODE_SERVER_URL__`
 * into index.html so the webgui can reach the opencode REST API.
 */
class WebguiStaticServer {
  private instances = new Map<string, { server: http.Server; base: string }>();

  async start(rootDir: string, opencodeServerUrl: string): Promise<string> {
    const root = path.resolve(rootDir);
    const serverUrl = opencodeServerUrl.replace(/\/$/, "");
    const key = `${root}|${serverUrl}`;
    const existing = this.instances.get(key);
    if (existing) {
      return existing.base;
    }

    return new Promise((resolve, reject) => {
      const server = http.createServer((req, res) => this.handle(req, res, root, serverUrl));
      server.listen(0, "127.0.0.1", () => {
        const addr = server.address();
        if (addr && typeof addr !== "string") {
          const base = `http://127.0.0.1:${addr.port}`;
          this.instances.set(key, { server, base });
          logger.appendLine(`WebguiStaticServer started on ${base} serving ${root}`);
          resolve(base);
        } else {
          server.close();
          reject(new Error("Failed to get server port"));
        }
      });
      server.on("error", (e) => {
        logger.appendLine(`WebguiStaticServer error: ${e}`);
        reject(e);
      });
    });
  }

  stop(baseUrl?: string): void {
    if (!baseUrl) {
      for (const item of this.instances.values()) {
        item.server.close();
      }
      this.instances.clear();
      return;
    }

    const normalized = baseUrl.replace(/\/$/, "");
    const entry = Array.from(this.instances.entries()).find(([, value]) => value.base === normalized);
    if (!entry) {
      return;
    }

    entry[1].server.close();
    this.instances.delete(entry[0]);
  }

  private handle(req: http.IncomingMessage, res: http.ServerResponse, rootDir: string, serverUrl: string): void {
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Access-Control-Allow-Methods", "GET, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");

    if (req.method === "OPTIONS") {
      res.writeHead(204);
      res.end();
      return;
    }

    const url = new URL(req.url || "/", "http://127.0.0.1");
    let pathname = decodeURIComponent(url.pathname);

    // Must start with /app/
    if (!pathname.startsWith("/app")) {
      // Redirect root to /app/
      if (pathname === "/") {
        res.writeHead(302, { Location: "/app/" });
        res.end();
        return;
      }
      res.writeHead(404);
      res.end("Not Found");
      return;
    }

    // Strip /app prefix to get the relative file path within webgui-dist
    let relative = pathname.slice("/app".length);
    if (!relative || relative === "/") {
      relative = "/index.html";
    }

    const root = path.resolve(rootDir);
    const file = path.resolve(root, `.${relative}`);

    // Prevent directory traversal
    if (file !== root && !file.startsWith(`${root}${path.sep}`)) {
      res.writeHead(403);
      res.end("Forbidden");
      return;
    }

    // Check if file exists
    if (!fs.existsSync(file) || fs.statSync(file).isDirectory()) {
      // SPA fallback: serve index.html for non-asset paths
      const index = path.join(root, "index.html");
      if (fs.existsSync(index)) {
        this.serveFile(res, index, true, serverUrl);
        return;
      }
      res.writeHead(404);
      res.end("Not Found");
      return;
    }

    this.serveFile(res, file, path.basename(file) === "index.html", serverUrl);
  }

  private serveFile(res: http.ServerResponse, file: string, inject: boolean, serverUrl: string): void {
    const ext = path.extname(file).toLowerCase();
    const mime = MIME[ext] || "application/octet-stream";

    if (inject && ext === ".html") {
      // Read, inject global, serve
      let html = fs.readFileSync(file, "utf-8");
      const script = `<script>window.__OPENCODE_SERVER_URL__=${JSON.stringify(serverUrl)};</script>`;
      // Insert before the first <script tag so the global is available when app code runs
      const idx = html.indexOf("<script");
      if (idx !== -1) {
        html = html.slice(0, idx) + script + "\n    " + html.slice(idx);
      } else {
        html = html.replace("</head>", `${script}\n</head>`);
      }
      res.writeHead(200, { "Content-Type": mime, "Cache-Control": "no-cache" });
      res.end(html);
      return;
    }

    // Stream the file
    res.writeHead(200, {
      "Content-Type": mime,
      "Cache-Control": ext === ".html" ? "no-cache" : "public, max-age=31536000, immutable",
    });
    fs.createReadStream(file).pipe(res);
  }
}

export const webguiServer = new WebguiStaticServer();

File: hosts/jetbrains-plugin/src/main/kotlin/paviko/opencode/ui/WebguiStaticServer.kt
package paviko.opencode.ui

import com.intellij.openapi.diagnostic.Logger
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.io.File
import java.net.InetSocketAddress
import java.net.URLDecoder
import java.util.concurrent.Executors

/**
 * Lightweight HTTP server that serves the embedded webgui-dist files
 * under the `/app/` prefix and injects `window.__OPENCODE_SERVER_URL__`
 * into index.html so the webgui can reach the opencode REST API.
 *
 * Mirrors hosts/vscode-plugin/src/ui/WebguiStaticServer.ts
 */
object WebguiStaticServer {
    private val LOG = Logger.getInstance(WebguiStaticServer::class.java)

    private data class Instance(val server: HttpServer, val base: String)
    private val instances = mutableMapOf<String, Instance>()

    private val MIME = mapOf(
        ".html" to "text/html; charset=utf-8",
        ".js" to "application/javascript; charset=utf-8",
        ".mjs" to "application/javascript; charset=utf-8",
        ".css" to "text/css; charset=utf-8",
        ".json" to "application/json; charset=utf-8",
        ".svg" to "image/svg+xml",
        ".png" to "image/png",
        ".jpg" to "image/jpeg",
        ".jpeg" to "image/jpeg",
        ".gif" to "image/gif",
        ".ico" to "image/x-icon",
        ".woff" to "font/woff",
        ".woff2" to "font/woff2",
        ".ttf" to "font/ttf",
        ".wasm" to "application/wasm",
        ".map" to "application/json",
    )

    @Synchronized
    fun start(root: String, opencodeServerUrl: String): String {
        val canonicalRoot = File(root).canonicalPath
        val normalizedServerUrl = opencodeServerUrl.trimEnd('/')
        val key = "$canonicalRoot|$normalizedServerUrl"

        val existing = instances[key]
        if (existing != null) return existing.base

        val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0).apply {
            executor = Executors.newCachedThreadPool()
            createContext("/") { exchange -> handle(exchange, canonicalRoot, normalizedServerUrl) }
            start()
        }
        val base = "http://127.0.0.1:${server.address.port}"
        instances[key] = Instance(server, base)
        LOG.info("WebguiStaticServer started on $base serving $canonicalRoot")
        return base
    }

    @Synchronized
    fun stop() {
        val current = instances.values.toList()
        instances.clear()
        current.forEach { it.server.stop(0) }
    }

    @Synchronized
    fun stop(baseUrl: String) {
        val normalizedBase = baseUrl.trimEnd('/')
        val target = instances.entries.firstOrNull { it.value.base == normalizedBase } ?: return
        target.value.server.stop(0)
        instances.remove(target.key)
    }

    private fun handle(exchange: HttpExchange, rootDir: String, serverUrl: String) {
        exchange.responseHeaders.apply {
            add("Access-Control-Allow-Origin", "*")
            add("Access-Control-Allow-Methods", "GET, OPTIONS")
            add("Access-Control-Allow-Headers", "Content-Type")
        }

        if (exchange.requestMethod == "OPTIONS") {
            exchange.sendResponseHeaders(204, -1)
            exchange.close()
            return
        }

        val pathname = try {
            URLDecoder.decode(exchange.requestURI.path ?: "/", "UTF-8")
        } catch (_: Throwable) {
            exchange.requestURI.path ?: "/"
        }

        // Must start with /app
        if (!pathname.startsWith("/app")) {
            if (pathname == "/") {
                exchange.responseHeaders.add("Location", "/app/")
                exchange.sendResponseHeaders(302, -1)
                exchange.close()
                return
            }
            exchange.sendResponseHeaders(404, -1)
            exchange.close()
            return
        }

        // Strip /app prefix to get relative file path within webgui-dist
        var relative = pathname.removePrefix("/app")
        if (relative.isEmpty() || relative == "/") relative = "/index.html"

        val file = File(rootDir, relative).canonicalFile
        val rootPath = rootDir

        // Prevent directory traversal
        if (file.path != rootPath && !file.path.startsWith("$rootPath${File.separator}")) {
            exchange.sendResponseHeaders(403, -1)
            exchange.close()
            return
        }

        // Check if file exists
        if (!file.exists() || file.isDirectory) {
            val leaf = relative.substringAfterLast('/')
            val isSpaRoute = !leaf.contains('.')
            if (isSpaRoute) {
                val index = File(rootDir, "index.html")
                if (index.exists()) {
                    serveFile(exchange, index, true, serverUrl)
                    return
                }
            }
            exchange.sendResponseHeaders(404, -1)
            exchange.close()
            return
        }

        serveFile(exchange, file, file.name == "index.html", serverUrl)
    }

    private fun serveFile(exchange: HttpExchange, file: File, inject: Boolean, serverUrl: String) {
        val ext = file.extension.let { if (it.isNotEmpty()) ".$it" else "" }.lowercase()
        val mime = MIME[ext] ?: "application/octet-stream"

        if (inject && ext == ".html") {
            var html = file.readText(Charsets.UTF_8)
            val escaped = serverUrl.replace("\"", "\\\"")
            val script = "<script>window.__OPENCODE_SERVER_URL__=\"$escaped\";</script>"
            val idx = html.indexOf("<script")
            html = if (idx != -1) {
                html.substring(0, idx) + script + "\n    " + html.substring(idx)
            } else {
                html.replace("</head>", "$script\n</head>")
            }
            val bytes = html.toByteArray(Charsets.UTF_8)
            exchange.responseHeaders.apply {
                add("Content-Type", mime)
                add("Cache-Control", "no-cache")
            }
            exchange.sendResponseHeaders(200, bytes.size.toLong())
            exchange.responseBody.use { it.write(bytes) }
            return
        }

        val bytes = file.readBytes()
        exchange.responseHeaders.apply {
            add("Content-Type", mime)
            add("Cache-Control", if (ext == ".html") "no-cache" else "public, max-age=31536000, immutable")
        }
        exchange.sendResponseHeaders(200, bytes.size.toLong())
        exchange.responseBody.use { it.write(bytes) }
    }
}
